### ISSUE TITLE
Bug: `overrideConfig` not working as expected.

### ISSUE THREAD

**Issue by fisker** [eslint/eslint#18563]
### Environment

Node version: 22.2.0
npm version: 10.7.0
Local ESLint version: 9.4.0
Global ESLint version: N/A
Operating System: Windows

### What parser are you using?

Default (Espree)

### What did you do?

```js
import { ESLint } from 'eslint';
const config = { rules: { quotes: 1 } };

console.log(
  await Promise.allSettled(
    [
      // 1
      new ESLint({ overrideConfigFile: true, overrideConfig: config }),

      // 2
      new ESLint({ overrideConfigFile: true, overrideConfig: [config] }),

      // 3
      new ESLint({ overrideConfigFile: true, overrideConfig: [[config]] }),
    ].map((eslint) => eslint.lintText(''))
  )
);
```

### What did you expect to happen?

1 should throw. 2, 3 should work. (Not sure if this is correct, it's unclear to me)

### What actually happened?

1, 2 works, 3 throws.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/stackblitz-starters-dctzrs?file=index.js

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

Problems:

1. It's not clear to me if the flat config can be an object or not, the RFC says [it can be an object](https://github.com/eslint/rfcs/blob/3d1f1d8dcd675a85dfc0542be9dac8c384a1369d/designs/2019-config-simplification/README.md#the-eslintconfigjs-file), but all [the examples](https://eslint.org/docs/latest/use/configure/configuration-files) uses array.

2. The array should be flatten, but it's not.

	> A config array is always flattened before being evaluated
	
	Quote from https://github.com/eslint/rfcs/blob/3d1f1d8dcd675a85dfc0542be9dac8c384a1369d/designs/2019-config-simplification/README.md#extending-another-config

I'm making a shareable config, not sure what format should I use.

1. Export an object (hard to add overrides in future)

	Can user

	```js
	export {default} from 'eslint-config-foo'
	```
	
	Or user have to

	```js
	import eslintConfigFoo from 'eslint-config-foo'

	export default [eslintConfigFoo]
	```

2. Export an array

	I'm sure this will work

	
	```js
	export {default} from 'eslint-config-foo'
	```

	But

	```js
	import eslintConfigFoo from 'eslint-config-foo'

	// Will this work?
	export default [eslintConfigFoo] 

	// Or user have to flat?
	
	export default [...eslintConfigFoo]
	export default [eslintConfigFoo].flat(Infinity)
	```

**Comment by mdjermanovic** [comment]
I believe this all works as intended, but we should probably update the documentation in a few places.

> ```js
>       // 1
>       new ESLint({ overrideConfigFile: true, overrideConfig: config }),
> 
>       // 2
>       new ESLint({ overrideConfigFile: true, overrideConfig: [config] }),
> ```

`overrideConfig` can be a config object or an array of config objects.

https://github.com/eslint/eslint/blob/518a35c8fa9161522cbe9066d48e6c6fcd8aadf3/lib/eslint/eslint.js#L444-L450

So it seems we should update the type in the docs to include arrays:

https://github.com/eslint/eslint/blob/518a35c8fa9161522cbe9066d48e6c6fcd8aadf3/docs/src/integrate/nodejs-api.md?plain=1#L145-L146


> ```js
> // 3
>       new ESLint({ overrideConfigFile: true, overrideConfig: [[config]] }),
> ```
> 2\. The array should be flatten, but it's not.
>     > A config array is always flattened before being evaluated
>     
>     
>     Quote from https://github.com/eslint/rfcs/blob/3d1f1d8dcd675a85dfc0542be9dac8c384a1369d/designs/2019-config-simplification/README.md#extending-another-config

While the RFC design mentions flattening, it's intentionally not implemented to make the use of shareable configurations simpler when they are applied to a subset of files only:

https://eslint.org/docs/latest/use/configure/combine-configs

> 1. It's not clear to me if the flat config can be an object or not, the RFC says [it can be an object](https://github.com/eslint/rfcs/blob/3d1f1d8dcd675a85dfc0542be9dac8c384a1369d/designs/2019-config-simplification/README.md#the-eslintconfigjs-file), but all [the examples](https://eslint.org/docs/latest/use/configure/configuration-files) uses array.

Config file can export a config object or an array of config objects:

https://github.com/eslint/eslint/blob/518a35c8fa9161522cbe9066d48e6c6fcd8aadf3/lib/eslint/eslint.js#L398-L402

All the examples indeed use arrays. I think we should keep it that way for simplicity, but maybe we could mention that config file can also export a single config object. @eslint/eslint-team thoughts?





> I'm making a shareable config, not sure what format should I use.
> 
> 1. Export an object (hard to add overrides in future)
>    Can user
>    ```js
>    export {default} from 'eslint-config-foo'
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    Or user have to
>    ```js
>    import eslintConfigFoo from 'eslint-config-foo'
>    
>    export default [eslintConfigFoo]
>    ```
> 2. Export an array
>    I'm sure this will work
>    ```js
>    export {default} from 'eslint-config-foo'
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    But
>    ```js
>    import eslintConfigFoo from 'eslint-config-foo'
>    
>    // Will this work?
>    export default [eslintConfigFoo] 
>    
>    // Or user have to flat?
>    
>    export default [...eslintConfigFoo]
>    export default [eslintConfigFoo].flat(Infinity)
>    ```

Shareable configs are programmatically used in config files, so it's up to the config authors to decide on the format as long as it is documented.

However, we recommend that shareable configs export either an array of config objects or a single config object, but not an array of arrays because it would make it complicated to apply the config to a subset of files only (https://eslint.org/docs/latest/use/configure/combine-configs).

https://eslint.org/docs/latest/extend/shareable-configs

**Comment by fisker** [comment]
> we recommend that shareable configs export either an array of config objects or a single config object,

If the config export an object, it will be hard to add overides in futures. For example:

```js
export default {
	rules: {foo: 'error'}
}
```

One day, we found that special files should have a different config, we have to change it to array (correct me if I'm wrong)

```js
export default [
	{
		rules: {foo: 'error'}
	},
	{
		files: "a-special-file",
		rules: {foo: 'off'}
	}
]
```

Then it will be a breaking change, since user uses

```js
export default [
	eslintConfigFoo,
	// ...
]
```

I hope we can reconsider to flat the array.

**Comment by aladdin-add** [comment]
in the new config, it's just a plain js module, so eslint no longer has the ability to enforce it.

you can always export an array (even if has only one config item):

```js
export default [config];
```

**Comment by fisker** [comment]
> eslint no longer has the ability to enforce it.

You are right ...

**Comment by mdjermanovic** [comment]
> One day, we found that special files should have a different config

I believe a change of this kind should by itself be a breaking change for a shareable config, but if you want to make it easier for users to upgrade, you could export an array from the start as @aladdin-add suggested, even though it contains only one item.

Then, you'd document that the exported config is an array and therefore should always be used like this:

```js
// eslint.config.js
import myconfig from "eslint-config-myconfig";

export default [
    ...myconfig
];
```

**Comment by fisker** [comment]
> you could export an array

I figured that.

But still, flatting will make things much easier.

If not acceptable,  I think we can also consider recommending shareable configs to use an array somewhere, or recommend user to use `.flat(Infinity)` , so user don't need check if it's an array or not.

**Comment by aladdin-add** [comment]
> But still, flatting will make things much easier.

I recall it was discussed somewhere, but not accepted.

**Comment by aladdin-add** [comment]
https://github.com/eslint/eslint/issues/18040

**Comment by fisker** [comment]
Thanks, but disappointing..

**Comment by aladdin-add** [comment]
closing as it's working as expected.
